support/ebpf/go_labels.ebpf.c

```diff
@@ -0,0 +1,266 @@
+      (unsigned long)record->labels[0].str,
+      res);
+    return;
+  }
+  unsigned vlen = MIN(record->labels[1].len, CUSTOM_LABEL_MAX_VAL_LEN - 1);
+  res           = bpf_probe_read_user(lbl->val, vlen, record->labels[1].str);
+  if (res) {
+    DEBUG_PRINT(
+      "cl: failed to read key for custom label (%lx): %ld",
+      (unsigned long)record->labels[1].str,
+      res);
+    return;
+  }
+  out->len++;
+}
+
+static inline __attribute__((__always_inline__)) bool
+get_go_custom_labels_from_slice(struct pt_regs *ctx, PerCPURecord *record, void *labels_slice_ptr)
+{
+  // https://github.com/golang/go/blob/80e2e474/src/runtime/pprof/label.go#L20
+  struct GoSlice labels_slice;
+  long res = bpf_probe_read_user(&labels_slice, sizeof(struct GoSlice), labels_slice_ptr);
+  if (res < 0) {
+    DEBUG_PRINT(
+      "cl: failed to read value for labels slice (%lx): %ld", (unsigned long)labels_slice_ptr, res);
+    return false;
+  }
+
+  u64 label_count = MIN(MAX_CUSTOM_LABELS, labels_slice.len);
+  switch (label_count) {
+  case 10: process_slice_pair(record, &labels_slice, 9);
+  case 9: process_slice_pair(record, &labels_slice, 8);
+  case 8: process_slice_pair(record, &labels_slice, 7);
+  case 7: process_slice_pair(record, &labels_slice, 6);
+  case 6: process_slice_pair(record, &labels_slice, 5);
+  case 5: process_slice_pair(record, &labels_slice, 4);
+  case 4: process_slice_pair(record, &labels_slice, 3);
+  case 3: process_slice_pair(record, &labels_slice, 2);
+  case 2: process_slice_pair(record, &labels_slice, 1);
+  case 1: process_slice_pair(record, &labels_slice, 0);
+  }
+
+  return true;
+}
+
+static inline __attribute__((__always_inline__)) bool get_go_custom_labels_from_map(
+  struct pt_regs *ctx, PerCPURecord *record, void *labels_map_ptr_ptr, GoCustomLabelsOffsets *offs)
+{
+  void *labels_map_ptr;
+  long res = bpf_probe_read_user(&labels_map_ptr, sizeof(labels_map_ptr), labels_map_ptr_ptr);
+  if (res < 0) {
+    DEBUG_PRINT(
+      "cl: failed to read value for labels_map_ptr (%lx): %ld",
+      (unsigned long)labels_map_ptr_ptr,
+      res);
+    return false;
+  }
+
+  u64 labels_count = 0;
+  res = bpf_probe_read_user(&labels_count, sizeof(labels_count), labels_map_ptr + offs->hmap_count);
+  if (res < 0) {
+    DEBUG_PRINT("cl: failed to read value for labels_count: %ld", res);
+    return false;
+  }
+  if (labels_count == 0) {
+    DEBUG_PRINT("cl: no labels");
+    return false;
+  }
+
+  unsigned char log_2_bucket_count;
+  res = bpf_probe_read_user(
+    &log_2_bucket_count, sizeof(log_2_bucket_count), labels_map_ptr + offs->hmap_log2_bucket_count);
+  if (res < 0) {
+    DEBUG_PRINT("cl: failed to read value for bucket_count: %ld", res);
+    return false;
+  }
+  void *label_buckets;
+  res =
+    bpf_probe_read_user(&label_buckets, sizeof(label_buckets), labels_map_ptr + offs->hmap_buckets);
+  if (res < 0) {
+    DEBUG_PRINT("cl: failed to read value for label_buckets: %ld", res);
+    return false;
+  }
```

Comment:
This does several reads from the `labels_map_ptr`. Reading once to a large scratch buffer would also work and probably be faster if the maximum size of the data to be read is not too large. A way to do this so that ebpf verifier still is happy is at:

https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/main/support/ebpf/hotspot_tracer.ebpf.c#L813-L839